_drafts/programatically-configured-bots.md

```diff
@@ -0,0 +1,95 @@
+
+## Motivation
+
+Hubot runs your code is by having it in a `scripts` directory, or specified in `external-scripts.json` (and the deprecated `hubot-scripts.json`). This means there is not an obvious place to inject framework-level customizations (like an install-specific `robot` helper function).
+
+Hubot loads scripts alphabetically in the script directory, so the current workaround is to make a script file named that Hubot will load first. Consider:
+
+```coffeescript
+# scripts/0-myawesomehelper.coffee
+module.exports = (robot) ->
+  robot.myawesomehelper = () ->
+    # insert awesome code here
+
+# scripts/whatever.coffee
+module.exports = (robot) ->
+  robot.myawesomehelper()
+```
+
+Some tools need to be initialized as soon as possible to instrument as much as possible. There's no easy way to load that any sooner than when Hubot starts loading scripts. [hubot#847](https://github.com/github/hubot/issues/847) is one example of this, where the user wants to instrument Hubot with [New Relic](https://newrelic.com/), but it can't be loaded soon enough to be effective.
+
+Hubot includes an instance of [Express](https://expressjs.com/) in it, for convenience in integrating with other services, and also because some adapters need it. Express is itself very configurable, and it's difficult to accommodate that configurability from Hubot. There have been PRs and requests over the year to allow customization of some of its Express, and have involved adding an increasing number of environment variables to control the behavior. An experienced Express user starting with Hubot would not be able to configure the server in all the ways they are familiar with. It's worth a separate evolution proposal for more specifics on that, but a programmatic launch of hubot would be a prerequisite of it.
+
+Lastly, there is a fair amount of scaffolding to get hubot running. Programmatic launching of hubot would allow for a single-file example hubot. This would make for a great getting started example, and also obviate the need for a generator which has been programmatic in the past.
+
+## Proposed solution
+
+> Describe your solution to the problem. Provide examples and describe how they work. Show how your solution is better than current workarounds.
+
+TODO
+
+```coffeescript
+Hubot   = require 'hubot'
+Log     = require 'log'
+Path    = require 'path'
+
+robot = new Hubot.Robot
+robot.logger = new Log('info')
+robot.adapter = new Hubot.BuiltinAdapters.Shell(robot)
+robot.alias = '/'
+
+# wait for adapter to connect, since it might change robot.name at runtime,
+# like robot name
+robot.adapter.on 'connected', ->
```

Comment:
This probably needs to be `once` instead of `on`, otherwise it'd register more listeners, and things would reply again.